### PR TITLE
Refactor lookupPath to avoid recursion. NFC

### PR DESCRIPTION
**Pull request by hoodmane**
This removes the recursion from `lookupPath` and in my opinion makes the control flow much more explicit and comprehensible.

### CODE REVIEW

File: src/library_fs.js [hunk 1]
```diff
@@ -177,57 +177,47 @@ FS.staticInit();
...
       opts = Object.assign(defaults, opts)
 
-      if (opts.recurse_count > 8) {  // max recursive lookup of 8
-        throw new FS.ErrnoError({{{ cDefs.ELOOP }}});
-      }
-
-      // split the absolute path
-      var parts = path.split('/').filter((p) => !!p);
+      // limit max consecutive symlinks to 40 (SYMLOOP_MAX).
+      linkloop: for (var nlinks = 0; nlinks < 40; nlinks ++) {
```
Comment (sbc100) [inline src/library_fs.js:184]: Wow, labeled `continue` targets!

Comment (sbc100) [inline src/library_fs.js:184]: No space before `++`

File: src/library_fs.js [hunk 2]
```diff
@@ -177,57 +177,47 @@ FS.staticInit();
 
       var defaults = {
         follow_mount: true,
-        recurse_count: 0
       };
       opts = Object.assign(defaults, opts)
```
Comment (sbc100) [inline src/library_fs.js:181]: Can we replace these 5 lines with `opts.follow_mount ??= true`?

**Review by sbc100 — COMMENTED** [review]
Thanks for working on this!

File: src/library_fs.js (hunk 2 above)
Comment (hoodmane) [reply src/library_fs.js]: I think so.

**Review by sbc100 — APPROVED** [review]
Can you rebase the codesize tests?

**Comment by hoodmane** [comment]
Well it's going to conflict with #22998, I suppose you can pick which one to merge first.

**Comment by sbc100** [comment]
Can you confirm those codesize changes are all due to this change?   i.e. can you run the rebase on `main` and confirm you don't see any changes?

**Comment by hoodmane** [comment]
There are a lot of changes when I do it on main too.

**Comment by sbc100** [comment]
> There are a lot of similar changes when I run it on main too.

Maybe try with `emsdk install tot` (or otherwise make sure you have the latest llvm and binaryen).   (Also try `./emcc --clear-cache`)

**Comment by hoodmane** [comment]
Even after `./emsdk install tot`, and clearing cache and getting the most up to date main branch I get a bunch of changes from the rebaseline.

**Comment by sbc100** [comment]
> Even after `./emsdk install tot`, and clearing cache and getting the most up to date main branch I get a bunch of changes from the rebaseline.

Hmm.. that is strange/annoying.  Maybe just revert those files then, and I can do a rebaseline after this lands.

**Comment by sbc100** [comment]
OK, I just pushed another rebaseline commit, perhaps this one will match what you are generated on main now?

**Comment by hoodmane** [comment]
Nope, even if I check out your rebaseline commit and do a clean install of tot from emsdk I get a bunch of changes in code size by one to ten bytes.

File: ChangeLog.md [hunk 3]
```diff
@@ -22,6 +22,9 @@ See docs/process.md for more on how version tagging works.
 -----------------------
 - libunwind was updated to LLVM 19.1.4. (#22394)
 - mimalloc was updated to 2.1.7. (#21548)
+- The file system was updated to independently track atime, mtime and ctime
+  instead of using the same time for all three.
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply ChangeLog.md:26]: Oh I think this is the wrong branch...

**Comment by hoodmane** [comment]
I think this PR fixes the case when you symlink a directory to itself `symlink(".", "dir")` and then try to `glob("**")` it. The previous `lookupPath` doesn't seem to detect this loop.

**Comment by hoodmane** [comment]
@sbc100 can this be merged?

**Comment by hoodmane** [comment]
Thanks!

**Pull request merged**